external/storm-kinesis/src/main/java/org/apache/storm/kinesis/spout/KinesisRecordsManager.java

```diff
@@ -0,0 +1,566 @@
+            } else if (shardIteratorType.equals(ShardIteratorType.AT_TIMESTAMP)) {
+                getShardIteratorRequest.setTimestamp(timestamp);
+            }
+            GetShardIteratorResult getShardIteratorResult = kinesisClient.getShardIterator(getShardIteratorRequest);
+            if (getShardIteratorResult != null) {
+                shardIterator = getShardIteratorResult.getShardIterator();
+            }
+        } catch (Exception e) {
+            LOG.debug("Exception occured while getting shardIterator for shard " + shardId + " shardIteratorType " + shardIteratorType + " sequence number " +
+                    sequenceNumber + " timestamp " + timestamp, e);
+        }
+        LOG.debug("Returning shardIterator " + shardIterator + " for shardId " + shardId + " shardIteratorType " + shardIteratorType + " sequenceNumber " +
+                sequenceNumber + " timestamp" + timestamp);
+        return shardIterator;
+    }
+
+    private Long getUncommittedRecordsCount () {
+        Long result = 0L;
+        for (Map.Entry<String, TreeSet<BigInteger>> emitted: emittedPerShard.entrySet()) {
+            result += emitted.getValue().size();
+        }
+        for (Map.Entry<String, TreeSet<BigInteger>> acked: ackedPerShard.entrySet()) {
+            result += acked.getValue().size();
+        }
+        for (Map.Entry<String, TreeSet<BigInteger>> failed: failedPerShard.entrySet()) {
+            result += failed.getValue().size();
+        }
+        LOG.debug("Returning uncommittedRecordsCount as " + result);
+        return result;
+    }
+
+    private boolean shouldFetchNewRecords () {
+        // check to see if any shard has already fetched records waiting to be emitted, in which case dont fetch more
+        boolean fetchRecords = true;
+        for (Map.Entry<String, LinkedList<Record>> entry: toEmitPerShard.entrySet()) {
+            if (!entry.getValue().isEmpty()) {
+                fetchRecords = false;
+                break;
+            }
+        }
+        return fetchRecords;
+    }
+
+    private void initializeCurator () {
+        ZkInfo zkInfo = config.getZkInfo();
+        curatorFramework = CuratorFrameworkFactory.newClient(zkInfo.getZkUrl(), zkInfo.getSessionTimeoutMs(), zkInfo.getConnectionTimeoutMs(), new
+                RetryNTimes(zkInfo.getRetryAttempts(), zkInfo.getRetryIntervalMs()));
+        curatorFramework.start();
+    }
+
+    private void initializeKinesisClient () {
+        kinesisClient = new AmazonKinesisClient(config.getKinesisConnectionInfo().getCredentialsProvider(), config.getKinesisConnectionInfo().getClientConfiguration());
+        kinesisClient.setRegion(Region.getRegion(config.getKinesisConnectionInfo().getRegion()));
+    }
+
+    private void shutdownCurator () {
+        curatorFramework.close();
+    }
+
+    private void shutdownKinesisClient () {
+        kinesisClient.shutdown();
+    }
+
+}
```

Comment:
This is big file thats doing lot of things. We should some of these methods into Utils class or something  like zk related methods into its own class.